# Profitable App Profiles for the App Store and Google Play Markets

Our aim in this project is to find mobile app profiles that are profitable for the App Store and Google Play markets. We're working as data analysts for a company that builds Android and iOS mobile apps, and our job is to enable our team of developers to make data-driven decisions with respect to the kind of apps they build.

At our company, we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. This means that our revenue for any given app is mostly influenced by the number of users that use our app. Our goal for this project is to analyze data to help our developers understand what kinds of apps are likely to attract more users.

## Opening and Exploring the Data

Collecting data for over 4 million apps requires a significant amount of time and money, so we'll try to analyze a sample of the data instead. To avoid spending resources on collecting new data ourselves, we should first try to see if we can find any relevant existing data at no cost. Luckily, these are two data sets that seem suitable for our goals:

- A [data set](https://www.kaggle.com/lava18/google-play-store-apps) containing data about approximately 10,000 Android apps from Google Play; the data was collected in August 2018. You can download the data set directly from [this link](https://dq-content.s3.amazonaws.com/350/googleplaystore.csv).
- A [data set](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps) containing data about approximately 7,000 iOS apps from the App Store; the data was collected in July 2017. You can download the data set directly from [this link](https://dq-content.s3.amazonaws.com/350/AppleStore.csv).

Let's start by opening and then continue with exploring these two data sets.

In [1]:
from csv import reader

# for apple store
opened_file = open('AppleStore.csv', encoding="utf8")
read_file = reader(opened_file)
ios_data = list(read_file)

# get the header separately
ios_header = ios_data[0]
#now exclude the header
ios_data = ios_data[1:]

# for android play store
opened_file = open('googleplaystore.csv', encoding="utf8")
read_file = reader(opened_file)
android_data = list(read_file)

# get the header separately
android_header = android_data[0]
#now exclude the header
android_data = android_data[1:]

Now, write a function `explore_data()` so that we'll read the both dataset in a easier way.

In [2]:
def explore_data(dataset, start, end, rows_and_columns = False):
    data_slice = dataset[start:end]
    
    for row in data_slice:
        print(row)
        print('\n')
    
    if rows_and_columns:
        print('Number of rows: ', len(dataset))
        print('Number of columns: ', len(dataset[0]))        

In [3]:
# checking for android
print(android_header)
print('\n')
explore_data(android_data, 0, 3, True)

print('\n')
print ('= = = = = = = = = = = = = = = = = = = = = = = = = = = =')
print('\n')

# checking for ios
print(ios_header)
print('\n')
explore_data(ios_data, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows:  10841
Number of columns:  13


= = = = = = = = = = = = = = = = = = = = = = = = = = = =


['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_dev

## Deleting Wrong Data
The Google Play data set has a dedicated [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion), and we can see that one of the [discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) outlines an error for row 10472. Let's print this row and compare it against the header and another row that is correct.

In [4]:
# printing the data and comparing it with another correct data row

# header
print(android_header)

# wrong data row
print(android_data[10472])

# any correct data row
print(android_data[0])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


The row 10472 corresponds to the app Life Made WI-Fi Touchscreen Photo Frame, The row is actually missing its "Category" value, which causes the column shift. As a consequence, we'll delete this row. 

In [5]:
# printing the length of the data list
print(len(android_data))

# don't run this more than once
del android_data[10472]

# printing the length again for checking purpose
print(len(android_data))

10841
10840


## Removing Duplicates Entries
### For Android Apps
### Part One

If we explore the Google Play data set long enough, we'll find that some apps have more than one entry.

In [6]:
# let's check for some famous apps
for app in android_data:
    app_name = app[0]
    if app_name == 'Facebook':
        print(app)
    elif app_name == 'Twitter':
        print(app)
    elif app_name == 'Instagram':
        print(app)

['Facebook', 'SOCIAL', '4.1', '78158306', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Twitter', 'NEWS_AND_MAGAZINES', '4.3', '11667403', 'Varies with device', '500,000,000+', 'Free', '0', 'Mature 17+', 'News & Magazines', 'August 6, 2018', 'Varies with device', 'Varies with device']
['Twitter', 'NEWS_AND_MAGAZINES', '4.3', '11667403', 'Varies with device', '500,000,000+', 'Free', '

Above we explored, that we've too many duplicated entries in our android_data. And may be if we explore ios_data, we should find out some duplications there too. So better if we first separate the unique apps and duplicate apps, this way we get the count of how many duplicated apps we've. 

In [7]:
# creating empty lists
android_unique_apps = []
android_duplicate_apps = []

# looping through android apps data
for app in android_data:
    # getting the app name from the first column
    app_name = app[0]
    
    # check if app is duplicate (appearing twice or more) - 
    # if its in unique app list, its mean the app is duplicated
    if app_name in android_unique_apps:
        # added into duplicate list
        android_duplicate_apps.append(app_name)
    else:
        # if not added to unique list
        android_unique_apps.append(app_name)
        
# printing the results
print('Number of Unique apps:', len(android_unique_apps))
print('\n')
print('Number of duplicate apps:', len(android_duplicate_apps))
print('\n')
print('Examples of duplicate apps:', android_duplicate_apps[:25])

Number of Unique apps: 9659


Number of duplicate apps: 1181


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software', 'MailChimp - Email, Marketing Automation', 'Crew - Free Messaging and Scheduling', 'Asana: organize team projects', 'Google Analytics', 'AdWords Express', 'Accounting App - Zoho Books', 'Invoice & Time Tracking - Zoho', 'join.me - Simple Meetings', 'Invoice 2go — Professional Invoices and Estimates', 'SignEasy | Sign and Fill PDF and other Documents']


We don't want to count certain apps more than once when we analyze data, so we need to remove the duplicate entries and keep only one entry per app. One thing we could do is remove the duplicate rows randomly, but we could probably find a better way.

If you examine the rows we printed two cells above for the Instagram app, the main difference happens on the fourth position of each row, which corresponds to the number of reviews. The different numbers show that the data was collected at different times. We can use this to build a criterion for keeping rows. We won't remove rows randomly, but rather we'll keep the rows that have the highest number of reviews because the higher the number of reviews, the more reliable the ratings.

To do that, we will:

- Create a dictionary where each key is a unique app name, and the value is the highest number of reviews of that app
- Use the dictionary to create a new data set, which will have only one entry per app (and we only select the apps with the highest number of reviews)

### Part Two

Building the dictionary:

In [8]:
# checking which index is of reviews
# print(android_data[77][3])

# initializing empty dictionary for story reviews for each app (only max one)
reviews_max = {}

# looping through android apps data
for app in android_data:
    # getting the app name from the first column
    app_name = app[0]
    
    # getting the app review from the fourth column
    app_review = float(app[3])
    
    # checking if app is already in dict and if it is, only store the maximum review entry into the dict against the app name
    if app_name in reviews_max and reviews_max[app_name] < app_review:
        reviews_max[app_name] = app_review
    
    # if not, simply put the review in dict against the app name
    elif app_review not in reviews_max:
        reviews_max[app_name] = app_review

In a previous code cell, we found that there are 1,181 cases where an app occurs more than once, so the length of our dictionary (of unique apps) should be equal to the difference between the length of our data set and 1,181.

In [9]:
print('Expected length of data list:', len(android_data) - len(android_duplicate_apps))
print('Actual length of dictionary:', len(reviews_max))

Expected length of data list: 9659
Actual length of dictionary: 9659


Now, let's use the reviews_max dictionary to remove the duplicates. For the duplicate cases, we'll only keep the entries with the highest number of reviews. In the code cell below:

- We start by initializing two empty lists, android_clean and already_added.
- We loop through the android data set, and for every iteration:
  - We isolate the name of the app and the number of reviews.
  - We add the current row (app) to the android_clean list, and the app name (name) to the already_added list if:
      - The number of reviews of the current app matches the number of reviews of that app as described in the reviews_max dictionary; and
      - The name of the app is not already in the already_added list. We need to add this supplementary condition to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry (for example, the Box app has three entries, and the number of reviews is the same). If we just check for reviews_max[name] == n_reviews, we'll still end up with duplicate entries for some apps.

In [10]:
# creating two lists - 
# first for separating the clearning data
# second to make sure the data is not repeated again
android_clean_data = []
already_added_data_android = []

# # looping through android apps data
for app in android_data:
    # getting the app name from the first column
    app_name = app[0]
    
    # getting the app review from the fourth column
    app_review = float(app[3])
    
    # checking if current review exists in our reviews_max dict -
    # and at the same time, checking data is not get repeated
    if(reviews_max[app_name] == app_review) and (app_name not in already_added_data_android):
        android_clean_data.append(app)
        already_added_data_android.append(app_name)

Now let's quickly explore the new data set, and confirm that the number of rows is 9,659.

In [25]:
explore_data(android_clean_data, 0, 5, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


Number of rows:  9659
Number of columns:  13


We have 9659 rows, just as expected.

## Removing Duplicates Entries
### For iOS Apps
### Part One

In [12]:
# checking ios data for duplications
ios_unique_apps = []
ios_duplicate_apps = []

for app in ios_data:
    app_name = app[2]
    
    if app_name not in ios_unique_apps:
        ios_unique_apps.append(app_name)
    else:
        ios_duplicate_apps.append(app)
        
print('Number of Actual apps: ', len(ios_data))
print('Number of Unique apps: ', len(ios_unique_apps))
print('Number of Duplicate apps: ', len(ios_duplicate_apps))
print('Duplicate apps names: ', ios_duplicate_apps[0], ios_duplicate_apps[1])

Number of Actual apps:  7197
Number of Unique apps:  7195
Number of Duplicate apps:  2
Duplicate apps names:  ['7579', '1089824278', 'VR Roller Coaster', '240964608', 'USD', '0', '67', '44', '3.5', '4', '0.81', '4+', 'Games', '38', '0', '1', '1'] ['10885', '1178454060', 'Mannequin Challenge', '59572224', 'USD', '0', '105', '58', '4', '4.5', '1.0.1', '4+', 'Games', '38', '5', '1', '1']


So we end up finding only two apps are repeating. Other than the data is completely fine and not required cleaning.

### Part Two
Building the dictionary:

In [13]:
# initializing empty dictionary for story reviews for each app (only max one)
rating_count_max = {}

# looping through android apps data
for app in ios_data:
    # getting the app name from the first column
    app_name = app[2]
    
    # getting the app review from the fourth column
    app_rating_count = float(app[6])
    
    # checking if app is already in dict and if it is, only store the maximum review entry into the dict against the app name
    if app_name in rating_count_max and rating_count_max[app_name] < app_rating_count:
        rating_count_max[app_name] = app_rating_count
    
    # if not, simply put the review in dict against the app name
    elif app_rating_count not in rating_count_max:
        rating_count_max[app_name] = app_rating_count

In [14]:
print('Expected length of data list:', len(ios_data) - len(ios_duplicate_apps))
print('Actual length of dictionary:', len(rating_count_max))

Expected length of data list: 7195
Actual length of dictionary: 7195


In [15]:
# creating two lists - 
# first for separating the clearning data
# second to make sure the data is not repeated again
ios_clean_data = []
already_added_data_ios = []

# # looping through android apps data
for app in ios_data:
    # getting the app name from the first column
    app_name = app[2]
    
    # getting the app review from the fourth column
    rating_count = float(app[6])
    
    # checking if current review exists in our reviews_max dict -
    # and at the same time, checking data is not get repeated
    if(rating_count_max[app_name] == rating_count) and (app_name not in already_added_data_ios):
        ios_clean_data.append(app)
        already_added_data_ios.append(app_name)

In [16]:
explore_data(ios_clean_data, 0, 5, True)

['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1']


['4', '282614216', 'eBay: Best App to Buy, Sell, Save! Online Shopping', '128512000', 'USD', '0', '262241', '649', '4', '4.5', '5.10.0', '12+', 'Shopping', '37', '5', '9', '1']


['5', '282935706', 'Bible', '92774400', 'USD', '0', '985920', '5320', '4.5', '5', '7.5.1', '4+', 'Reference', '37', '5', '45', '1']


Number of rows:  7195
Number of columns:  17


We have 7195 rows, just as expected.

## Removing Non-English Apps
### Part One

If you explore the data sets enough, you'll notice the names of some of the apps suggest they are not directed toward an English-speaking audience. Below, we see a couple of examples from both data sets:

In [29]:
print(ios_clean_data[814][2])

print(android_clean_data[3046][0])

搜狐新闻—新闻热点资讯掌上阅读软件
صور حرف H


We're not interested in keeping these kind of apps, so we'll remove them. One way to go about this is to remove each app whose name contains a symbol that is not commonly used in English text — English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;, etc.), and other symbols (+, *, /, etc.).

All these characters that are specific to English texts are encoded using the ASCII standard. Each ASCII character has a corresponding number between 0 and 127 associated with it, and we can take advantage of that to build a function that checks an app name and tells us whether it contains non-ASCII characters.

We built this function below, and we use the built-in ord() function to find out the corresponding encoding number of each character.

In [30]:
def is_english(str):
    for char in str:
        if ord(char) > 127:
            return False
        
    return True

In [32]:
print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
False
False


The function seems to work fine, but some English app names use emojis or other symbols (™, — (em dash), – (en dash), etc.) that fall outside of the ASCII range. Because of this, we'll remove useful apps if we use the function in its current form.

In [33]:
def is_english(str):
    non_ascii = 0
    
    for char in str:
        if ord(char) > 127:
            non_ascii += 1
            
    if non_ascii > 3:
        return False
    else:
        return True
    
print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
True
True


The function is still not perfect, and very few non-English apps might get past our filter, but this seems good enough at this point in our analysis — we shouldn't spend too much time on optimization at this point.

Below, we use the is_english() function to filter out the non-English apps for both data sets:

In [34]:
android_data_english = []
ios_data_english = []

for app in android_clean_data:
    app_name = app[0]
    
    if is_english(app_name):
        android_data_english.append(app)
    
for app in ios_clean_data:
    app_name = app[2]
    
    if is_english(app_name):
        ios_data_english.append(app)
        
explore_data(android_data_english, 0, 5, True)
print('\n\n')
explore_data(ios_data_english, 0, 5, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


Number of rows:  9614
Number of columns:  13



['1', '281656475', 'PAC-MAN Premium', '1007

We can see that we're left with 9614 Android apps and 6183 iOS apps.

## Isolating the Free Apps
As we mentioned in the introduction, we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. Our data sets contain both free and non-free apps, and we'll need to isolate only the free apps for our analysis. Below, we isolate the free apps for both our data sets.

In [36]:
android_data_final = []
ios_data_final = []

for app in android_data_english:
    app_price = app[7]
    
    if app_price == '0':
        android_data_final.append(app)
        
for app in ios_data_english:
    app_price = app[5]
    
    if app_price == '0.0':
        android_data_final.append(app)
        
print(len(android_data_final))
print(len(ios_data_final))

8864
0
